lib/runTx.js

```diff
@@ -148,53 +147,50 @@ module.exports = function (opts, cb) {
       }
 
       results.amountSpent = results.gasUsed.mul(new BN(tx.gasPrice))
-      // refund the leftover gas amount
-      fromAccount.balance = new BN(tx.gasLimit).sub(results.gasUsed)
-        .mul(new BN(tx.gasPrice))
-        .add(new BN(fromAccount.balance))
-
-      self.stateManager.cache.put(tx.from, fromAccount)
-      self.stateManager.touched.push(tx.from)
-
-      var minerAccount = self.stateManager.cache.get(block.header.coinbase)
-      // add the amount spent on gas to the miner's account
-      minerAccount.balance = new BN(minerAccount.balance)
-        .add(results.amountSpent)
-
-      // save the miner's account
-      if (!(new BN(minerAccount.balance).isZero())) {
-        self.stateManager.cache.put(block.header.coinbase, minerAccount)
+
+      async.series([
+        updateFromAccount,
+        updateMinerAccount,
+        cleanupAccounts
+      ], cb)
+
+      function updateFromAccount (next) {
+        // refund the leftover gas amount
+        var finalFromBalance = new BN(tx.gasLimit).sub(results.gasUsed)
+          .mul(new BN(tx.gasPrice))
+          .add(new BN(fromAccount.balance))
+        fromAccount.balance = finalFromBalance
+
+        self.stateManager.putAccountBalance(utils.toBuffer(tx.from), finalFromBalance, next)
       }
 
-      if (!results.vm.selfdestruct) {
-        results.vm.selfdestruct = {}
+      function updateMinerAccount (next) {
+        var minerAccount = self.stateManager.cache.get(block.header.coinbase)
+        // add the amount spent on gas to the miner's account
+        minerAccount.balance = new BN(minerAccount.balance)
+          .add(results.amountSpent)
+
+        // save the miner's account
+        if (!(new BN(minerAccount.balance).isZero())) {
+          self.stateManager.cache.put(block.header.coinbase, minerAccount)
```

Comment:
Is there a reason that this stays on acting on the cache directly and not switch to ``putAccount`` or ``putAccountBalance``?

Reply:
Not that I can think of. I was just refactoring and didn't change it. We should probably have a separate ticket to clean up all these direct cache writes